runtime/autoload/man.vim

```diff
@@ -11,127 +22,207 @@ catch /E145:/
-  let s:man_tag_depth = s:man_tag_depth + 1
+" save current position in the stack
+function! s:push_tag() abort
+  let s:tag_stack += [{
+        \ 'buf':  bufnr('%'),
+        \ 'lnum': line('.'),
+        \ 'col':  col('.'),
+        \ }]
+endfunction
 
-  let editcmd = 'edit'
-  " Use an existing 'man' window, else open a new one.
-  if &filetype !=# 'man'
+" find the closest man window above/left
+function! s:find_man(cmd) abort
+  if g:man_find_window != 1 || &filetype ==# 'man'
+    return a:cmd
+  endif
+  if winnr('$') > 1
     let thiswin = winnr()
-    wincmd b
-    if winnr() > 1
-      exec thiswin . 'wincmd w'
-      while 1
-        if &filetype ==# 'man'
-          break
-        endif
-        wincmd w
-        if thiswin == winnr()
-          break
-        endif
-      endwhile
-    endif
+    while 1
+      if &filetype ==# 'man'
+        return 'edit'
+      endif
+      wincmd w
+      if thiswin == winnr()
+        return a:cmd
+      endif
+    endwhile
+  endif
+  return a:cmd
+endfunction
 
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
-    endif
+" parses the page and sect out of 'page(sect)'
+function! s:parse_page_and_sect_fpage(fpage) abort
+  let ret = split(a:fpage, '(')
+  if len(ret) > 1
+    let iret = split(ret[1], ')')
+    return [ret[0], tolower(iret[0])]
+  else
+    return [ret[0], '']
   endif
+endfunction
 
-  silent exec editcmd.' man://'.page.(empty(sect)?'':'('.sect.')')
+" parses the page and sect out of 'path/page.sect'
+function! s:parse_page_and_sect_path(path) abort
+  let tail = fnamemodify(a:path, ':t')
+  if fnamemodify(tail, ':e') =~# s:man_extensions
+    let tail = fnamemodify(tail, ':r')
+  endif
+  " TODO(nhooyr) all substitutes to matchlist/matchstr maybe?
+  let page = substitute(tail, '^\(\f\+\)\..\+$', '\1', '')
+  let sect = substitute(tail, '^\f\+\.\(.\+\)$', '\1', '')
+  return [page, sect]
+endfunction
 
+function! s:read_page(sect, page, cmd)
+  silent execute s:find_man(a:cmd) 'man://'.a:page.(empty(a:sect)?'':'('.a:sect.')')
   setlocal modifiable
-  silent keepjumps norm! 1G"_dG
-  if empty($MANWIDTH)
-    let $MANWIDTH = winwidth(0)
-  endif
-  silent exec 'r!/usr/bin/man '.s:cmd(sect, page).' | col -b'
-  " Remove blank lines from top and bottom.
+  " remove all the text, incase we already loaded the manpage before
+  silent keepjumps %delete _
+  let $MANWIDTH = winwidth(0)-1
+  " read manpage into buffer
+  silent execute 'r!'.s:man_cmd.s:man_args(a:sect, a:page)
+  call man#normalize_page()
+  setlocal filetype=man
+endfunction
+
+function! man#normalize_page()
+  " remove all those backspaces
+  execute "silent keepjumps %substitute,.\b,,g"
+  " remove blank lines from top and bottom.
   while getline(1) =~# '^\s*$'
     silent keepjumps 1delete _
   endwhile
+  " TODO(nhooyr) is deleting the bottom lines necessary?
+  " I think only deleting the first line is necessary when using r! to
+  " read manpage in.
```

Comment:
Is deleting the bottom lines necessary? I think only deleting the first line is necessary when using r! to read manpage in.